src/mmt_tcpip/lib/protocols/http2.c

```diff
@@ -267,3 +267,201 @@ int mmt_check_http2(ipacket_t *ipacket, unsigned proto_index) {
+   	 }
+   	 new_length=new_length & 0x00FFFFFF;
+   	// printf("new_length %d, header_length %d\n",new_length,header_length);
+   	// printf("data_out %02X packet_data %02X \n",data_out[proto_offset],packet->data[proto_offset]);
+ 	memcpy((uint8_t*)data_out+proto_offset,packet->data+proto_offset,new_length+9);
+ 	return new_length-header_length;
+}
+int modify_get(uint8_t*data_out,int proto_offset){
+    	//Go to method field
+   	int header_length =0;
+      	// Get http2 protocol offset
+   	int offset_header_length = proto_offset -1;
+   	for (int i = offset_header_length; i < offset_header_length+4; i++) {
+      		  header_length = (header_length << 8) | data_out[i];
+   	 }
+   	header_length=header_length & 0x00FFFFFF;
+	uint8_t authority_amf[] = {0x41,0x8d,0x0b,0xa2,0x5c,0x2e,0x2e,0xdb,0xeb,0xba,0xcd,0xc7,0x80,0xf0,0x3f,0x7a,0x03,0x61,0x6d,0x66};
+	int authority_amf_length = sizeof(authority_amf) ;
+	//printf("[modify_get]authority_amf_length %d\n",authority_amf_length);
+	memcpy(data_out+proto_offset+9+header_length-2,authority_amf,authority_amf_length);
+	//printf("modify get after update:\n");
+	//for(int i=proto_offset+9+header_length-2;i<proto_offset+9+header_length-2+authority_amf_length;i++)
+	//	printf("%02hhX ",data_out[i]);
+	header_length+=authority_amf_length-2;//Http2 buffers the authority and amf with 2 bytes(0xc0 0xc1)from the first packet.The code restore the first authority_amf value.
+	//The new header lenght will be the sizeof authority and amf minus 2
+	data_out[offset_header_length+1]=header_length>>16; //I update the last 24 bites
+	data_out[offset_header_length+2]=header_length>>9; //I update the last 16 bites
+	data_out[offset_header_length+3]=header_length ;//I update the last 8 bites
+	//printf("[modify_get]offset_header_length %02hhX %02hhX %02hhX \n",data_out[offset_header_length+1],data_out[offset_header_length+2],data_out[offset_header_length+3]);
+		return (int)(sizeof(authority_amf)-2);
+}
+uint32_t update_window_update(char *data_out,int proto_offset,uint32_t modify){
+	int window_size_offset=proto_offset+9;
+	//int offset_header_length = proto_offset -1;
+	if(modify==1){
+		data_out[window_size_offset]=0x00;
+		data_out[window_size_offset+1]=0x00;
+		data_out[window_size_offset+2]=0x00;
+		data_out[window_size_offset+3]=0xFF;	
+		int difference_size=-9;
+		return difference_size;
+	}
+	//int header_length=0;
+   	//int header_length =ntohl( *((unsigned int *) & packet->data[offset_header_length]));
+	return 0;
+}
+
+int inject_http2_packet(uint8_t*data_out, uint8_t*data_to_inject,int proto_offset,int data_to_inject_len){
+   	int offset_header_length = proto_offset -1;
+   	int header_length=0;
+   	for (int i = offset_header_length; i < offset_header_length+4; i++) {
+      		  header_length = (header_length << 8) | data_out[i];
+   	 }
+   	// printf("inject_http2_packet Header_length %d",header_length);
+ 	memcpy(data_out+proto_offset,data_to_inject,data_to_inject_len);
```

Comment:
Before doing `memcpy`, you need to check whether `data_to_inject_len` is not bigger than the capability of `data_out+proto_offset`